heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCacheQueryUtils.java

```diff
@@ -0,0 +1,290 @@
+                TopologyMaster.MetricResponse.IndividualMetric.IntervalValue.newBuilder();
+
+            builder3.setValue(v.value);
+            builder3.setInterval(TopologyMaster.MetricInterval.newBuilder()
+                .setStart(v.startTime).setEnd(v.endTime));
+
+            builder2.addIntervalValues(builder3);
+          }// end IntervalValue
+        }
+
+        builder1.addMetric(builder2);
+      }// end IndividualMetric
+
+      builder.addMetric(builder1);
+    }// end TaskMetric
+
+    return builder.build();
+  }
+
+  // compatible with com.twitter.heron.proto.tmaster.TopologyMaster.ExceptionLogRequest
+  public static ExceptionRequest Convert(TopologyMaster.ExceptionLogRequest request) {
+    String componentName = request.getComponentName();
+
+    ExceptionRequest request1 = new ExceptionRequest();
+    request1.componentNameInstanceId = new HashMap<>();
+    Set<String> instances = new HashSet<>();
+    // only one component
+    request1.componentNameInstanceId.put(componentName, instances);
+
+    if (request.getInstancesCount() > 0) {
+      instances.addAll(request.getInstancesList());
+    } // empty list means all instances
+
+    return request1;
+  }
+
+  // compatible with com.twitter.heron.proto.tmaster.TopologyMaster.ExceptionLogResponse
+  public static TopologyMaster.ExceptionLogResponse Convert(ExceptionResponse response) {
+    TopologyMaster.ExceptionLogResponse.Builder builder =
+        TopologyMaster.ExceptionLogResponse.newBuilder();
+    // default OK if we have response to build already
+    builder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.OK));
+
+    for (ExceptionDatapoint e : response.exceptionDatapointList) {
+      TopologyMaster.TmasterExceptionLog.Builder builder1 =
+          TopologyMaster.TmasterExceptionLog.newBuilder();
+      // ExceptionDatapoint
+      builder1.setComponentName(e.componentName);
+      builder1.setHostname(e.hostname);
+      builder1.setInstanceId(e.instanceId);
+      // ExceptionData
+      builder1.setStacktrace(e.stacktrace);
+      builder1.setLasttime(e.lasttime);
+      builder1.setFirsttime(e.firsttime);
+      builder1.setCount(e.count);
+      builder1.setLogging(e.logging);
+
+      builder.addExceptions(builder1);
+    }
+
+    return builder.build();
+  }
+
+
+  public static class ExceptionRequest {
+    public Map<String, Set<String>> componentNameInstanceId;
+  }
+
+  public static class ExceptionResponse {
+    public List<ExceptionDatapoint> exceptionDatapointList;
+  }
+
+
+  public static class MetricTimeRangeValue {
+    public long startTime;
+    public long endTime;
+    public String value;
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("[")
+          .append(startTime).append("-").append(endTime)
+          .append(":")
+          .append(value)
+          .append("]");
+      return sb.toString();
+    }
+  }
+
+  public static class MetricDatum {
```

Comment:
For each of these model objects, can we make them immutable with a getter to get the values and all values set via constructor? IDEA makes it easy to generate that code. Typically we want to favor immutability. Same for all other model objects below.